usdigitalresponse/grants-ingest
# Feature: Persist forecasted grant opportunities to DynamoDB

### Ticket #861 

## Description

This PR modifies the `PersistGrantsGovXMLDB` pipeline step so that it supports persisting forecasted grant opportunities to DynamoDB. The process of taking XML data from an S3 object and persisting it to a DynamoDB item remains basically the same, except that all new records are persisted with an `is_forecast` boolean field, which can be used to discriminate between `OpportunitySynopsisDetail_1_0` (`is_forecast=false`) and `OpportunityForecastDetail_1_0` (`is_forecast=true`) source XML record types.

Summary of changes:
- A new `grantRecord` interface type is defined, which is implemented by both `opportunity` and `forecast` struct types.
- XML unmarshaling now uses an `xml.Decoder` instead of calling `xml.Unmarshal()`, in order to explicitly map `OpportunitySynopsisDetail_1_0` and `OpportunityForecastDetail_1_0` elements to their respective Go types.
- The `dynamodb.go` helper functions have been tweaked to only deal with types provided by the DynamoDB SDK rather than custom types defined locally, which makes it easier and cleaner to write unit tests for this file.
- Updated unit tests to reflect changes, with particular focus on ensuring that records are saved to DynamoDB with the correct `is_forecast` attribute value.
- Renamed custom Datadog metrics to use more generic `"record"` terminology instead of `"opportunity"`
- Added an S3-to-lambda bucket notification so that `PersistGrantsGovXMLDB` is invoked when forecasted grant XML records are uploaded to S3. Note that this has no effect unless forecasted grants are being uploaded by `SplitGrantsGovXMLDB`.

## Testing

- Start LocalStack using the included `docker-compose.yml` file
- Deploy to LocalStack (hint: run Taskfile command `task local:from-scratch`)
- Kick off the pipeline using the steps in README (hint: run `task local:invoke-DownloadGrantsGovDB`)
- Check that all records saved to DynamoDB have an `is_forecast` attribute with the correct value

### Automated and Unit Tests

### Manual tests for Reviewer

## Checklist
